Fix an infinite loop when using `#fmt: on/off` in the middle of an expression or code block

yilei opened the pull request:
<!-- Hello! Thanks for submitting a PR. To help make things go a bit more
     smoothly we would appreciate that you go through this template. -->

### Description

Fixes an infinite loop when using `#fmt: on/off` in the middle of an expression or code block.

This:

- resolves #3129
- resolves #3026
- resolves #2015

### Checklist - did you ...

<!-- If any of the following items aren't relevant for your contribution
     please still tick them so we know you've gone through the checklist.

    All user-facing changes should get an entry. Otherwise, signal to us
    this should get the magical label to silence the CHANGELOG entry check.
    Tests are required for bugfixes and new features. Documentation changes
    are necessary for formatting and most enhancement changes. -->

- [x] Add a CHANGELOG entry if necessary?
- [x] Add / update tests if necessary?
- [x] Add new / update outdated documentation?

<!-- Just as a reminder, everyone in all psf/black spaces including PRs
     must follow the PSF Code of Conduct (link below).

     Finally, once again thanks for your time and effort. If you have any
     feedback in regards to your experience contributing here, please
     let us know!

     Helpful links:

      PSF COC: https://www.python.org/psf/conduct/
      Contributing docs: https://black.readthedocs.io/en/latest/contributing/index.html
      Chat on Python Discord: https://discord.gg/RtVdv86PrH -->

ichard26 reviewed (commented):
The `# fmt: off/on/skip` might be the most hacky and impossibly-to-maintain complex feature of Black. Writing any sort of robust code here is challenging. This is definitely a step forward, but there's a wart that I'm not sure how to resolve.

My suggestion is to just fix the weird formatting issue as it involves a (IMO) valid use of `# fmt: off/on` and let the other usages break since they don't use it properly`. OTOH, there might be cases with valid usages of `# fmt: off/on` that break by fixing the formatting issue with #2015's example ...

yilei commented:
> `# fmt: off/on/skip` might be the most hacky, impossibly-hard-to-maintain, and complex feature of Black

The `STANDALONE_COMMENT` a neat trick though :)

Thanks for the review and suggestion!

I updated the code to your suggested "if child is a Leaf (so doesn't contain children) and contains a # fmt: on, then return", so the weird `+ path,` formatting is fixed.

I also added another hack to include this child Leaf if this is a closing bracket (it's an invalid use), so it doesn't fail for the first example.

Please take another look.

yilei commented on tests/data/simple_cases/fmtonoff5.py:
```diff
@@ -0,0 +1,37 @@
+    ]
+    # fmt: on
+    + path
+    ,
```
Updated so this is now fixed.

ichard26 reviewed (approved):
SGTM. Thank you for tackling this bug! 

P.S. you need to add the closing keyword before each issue you'd like to close automagically 
![image](https://user-images.githubusercontent.com/63936253/178545025-9825549c-2ce9-48e5-957e-90d952deecec.png)

yilei commented:
thanks for the tip! i updated this PR (and #3148)'s description to use the correct syntax.
